lib/l1_watcher/src/watcher.rs

```diff
@@ -3,64 +3,68 @@ use alloy::primitives::BlockNumber;
 
-pub(crate) struct L1Watcher<Event> {
+pub struct L1Watcher<Processor> {
     zk_chain: ZkChain<DynProvider>,
     next_l1_block: BlockNumber,
     max_blocks_to_process: u64,
-    _event: PhantomData<Event>,
+    poll_interval: Duration,
+    processor: Processor,
 }
 
-#[derive(Debug, Clone)]
-pub(crate) struct L1WatcherOutput<Event> {
-    /// New events that have been discovered since the last poll.
-    pub(crate) events: Vec<Event>,
-    /// Whether there are more blocks to process.
-    pub(crate) more_available: bool,
-}
-
-impl<Event> L1Watcher<Event> {
+impl<Processor: ProcessL1Event> L1Watcher<Processor> {
     pub(crate) fn new(
         zk_chain: ZkChain<DynProvider>,
         next_l1_block: BlockNumber,
         max_blocks_to_process: u64,
+        poll_interval: Duration,
+        processor: Processor,
     ) -> Self {
         Self {
             zk_chain,
             next_l1_block,
             max_blocks_to_process,
-            _event: PhantomData,
+            poll_interval,
+            processor,
         }
     }
 }
 
-impl<Event: WatchedEvent> L1Watcher<Event> {
-    /// Scans up to `self.max_blocks_to_process` next L1 blocks for new events of type `Event`
-    /// and returns them.
-    pub(crate) async fn poll(
-        &mut self,
-    ) -> Result<L1WatcherOutput<Event>, L1WatcherError<Event::Error>> {
-        let latest_block = self.zk_chain.provider().get_block_number().await?;
-        let from_block = self.next_l1_block;
-        // Inspect up to `self.max_blocks_to_process` blocks at a time
-        let to_block = latest_block.min(from_block + self.max_blocks_to_process - 1);
-        if from_block > to_block {
-            return Ok(L1WatcherOutput {
-                events: vec![],
-                more_available: false,
-            });
+impl<Processor: ProcessL1Event> L1Watcher<Processor> {
+    pub async fn run(mut self) -> Result<(), L1WatcherError<Processor::Error>> {
+        let mut timer = tokio::time::interval(self.poll_interval);
+        loop {
+            timer.tick().await;
+            self.poll().await?;
         }
-        let events = self.process_l1_blocks(from_block, to_block).await?;
-        METRICS.events_loaded[&Event::NAME].inc_by(events.len() as u64);
-        METRICS.most_recently_scanned_l1_block[&Event::NAME].set(to_block);
+    }
+
+    async fn poll(&mut self) -> Result<(), L1WatcherError<Processor::Error>> {
+        // Proactively iterate while there are more blocks to process.
+        loop {
+            let latest_block = self.zk_chain.provider().get_block_number().await?;
+            let from_block = self.next_l1_block;
+            // Inspect up to `self.max_blocks_to_process` blocks at a time
+            let to_block = latest_block.min(from_block + self.max_blocks_to_process - 1);
+            if from_block > to_block {
+                return Ok(());
+            }
+            let events = self.process_l1_blocks(from_block, to_block).await?;
```

Comment:
nit: seems like we use the term `process` for two different meanings - processing l1 blocks (this method), but also `process_event(event)` - that is, individual delegation to the concrete processor. Not a big deal, but a little confusing that actual `process`ing doesn't happen inside `process_l1_blocks` method.